File: code/7._Lempel-Ziv-Welch(LZW)/C#/programming.cs/LZW.cs
using System;
using System.Collections.Generic;

public class LZWCompression
{
    public static List<int> Compress(string input)
    {
        Dictionary<string, int> dictionary = new Dictionary<string, int>();
        for (int i = 0; i < 256; i++)
        {
            dictionary[Convert.ToChar(i).ToString()] = i;
        }

        List<int> result = new List<int>();
        int start = 0;
        int currentCode = 256;

        for (int i = 1; i <= input.Length; i++)
        {
            string currentSubstring = input.Substring(start, i - start);

            if (!dictionary.ContainsKey(currentSubstring) && i < input.Length)
            {
                dictionary[currentSubstring] = currentCode;
                currentCode++;
            }
            else
            {
                if (i == input.Length)
                {
                    result.Add(dictionary[currentSubstring]);
                }
                start = i;
            }
        }

        return result;
    }

    public static string Decompress(List<int> compressedData)
    {
        Dictionary<int, string> dictionary = new Dictionary<int, string>();
        for (int i = 0; i < 256; i++)
        {
            dictionary[i] = Convert.ToChar(i).ToString();
        }

        List<string> result = new List<string>();
        result.Add(dictionary[compressedData[0]]);
        int currentCode = 256;

        for (int i = 1; i < compressedData.Count; i++)
        {
            int code = compressedData[i];
            string entry;

            if (dictionary.ContainsKey(code))
            {
                entry = dictionary[code];
            }
            else if (code == currentCode)
            {
                entry = result[0];
            }
            else
            {
                throw new ArgumentException("Corrupted data");
            }

            result.Add(entry);
            dictionary[currentCode] = result[0];
            currentCode++;
        }

        return string.Join("", result);
    }

    static void Main()
    {
        string originalText = "ABABABABA";
        List<int> compressedData = Compress(originalText);
        Console.Write("Compressed Data: ");
        foreach (int code in compressedData)
        {
            Console.Write(code + " ");
        }
        Console.WriteLine();
        string decompressedText = Decompress(compressedData);
        Console.WriteLine("Decompressed Text: " + decompressedText);
    }
}

File: code/7._Lempel-Ziv-Welch(LZW)/JavaScript/programming.js/LZW.js
function compress(input) {
  const dictionary = {};
  for (let i = 0; i < 256; i++) {
    dictionary[String.fromCharCode(i)] = i;
  }

  const result = [];
  let start = 0;
  let currentCode = 256;

  for (let i = 1; i <= input.length; i++) {
    const currentSubstring = input.substring(start, i);

    if (!dictionary[currentSubstring] && i < input.length) {
      dictionary[currentSubstring] = currentCode;
      currentCode++;
    } else {
      if (i === input.length) {
        result.push(dictionary[currentSubstring]);
      }
      start = i;
    }
  }

  return result;
}

function decompress(compressedData) {
  const dictionary = {};
  for (let i = 0; i < 256; i++) {
    dictionary[i] = String.fromCharCode(i);
  }

  const result = [dictionary[compressedData[0]]];
  let currentCode = 256;

  for (let i = 1; i < compressedData.length; i++) {
    const code = compressedData[i];
    let entry = "";

    if (dictionary[code]) {
      entry = dictionary[code];
    } else if (code === currentCode) {
      entry = result[0];
    } else {
      throw new Error("Corrupted data");
    }

    result.push(entry);
    dictionary[currentCode] = result[0];
    currentCode++;
  }

  return result.join("");
}

const originalText = "ABABABABA";
const compressedData = compress(originalText);
console.log("Compressed Data: " + compressedData.join(" "));
const decompressedText = decompress(compressedData);
console.log("Decompressed Text: " + decompressedText);
